## Data Processing

In [2]:
# import libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [3]:
# import dataset

train_data = pd.read_csv('data/training.csv')
test_data = pd.read_csv('data/testing.csv')

In [4]:
train_data.head()

itching  skin_rash  continuous_sneezing  shivering  chills  joint_pain  \
0        0          1                    0          0       0           0   
1        0          1                    0          0       0           0   
2        0          0                    0          0       0           0   
3        0          1                    0          0       0           0   
4        0          1                    0          0       0           0   

   stomach_pain  acidity  ulcers_on_tongue  muscle_wasting  ...  \
0             0        0                 0               0  ...   
1             0        0                 0               0  ...   
2             0        0                 0               0  ...   
3             0        0                 0               0  ...   
4             0        0                 0               0  ...   

   red_spots_over_body  watering_from_eyes  increased_appetite  polyuria  \
0                    0                   0                   0         0   
1                    0                   0                   0         0   
2                    0                   0                   0         0   
3                    0                   0                   0         0   
4                    0                   0                   0         0   

   lack_of_concentration  visual_disturbances  pus_filled_pimples  blackheads  \
0                      0                    0                   1           1   
1                      0                    0                   1           1   
2                      0                    0                   1           1   
3                      0                    0                   0           1   
4                      0                    0                   1           0   

   scurring  prognosis  
0         1          0  
1         1          0  
2         1          0  
3         1          0  
4         1          0  

[5 rows x 58 columns]

In [6]:
train_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1440 entries, 0 to 1439
Data columns (total 58 columns):
 #   Column                        Non-Null Count  Dtype
---  ------                        --------------  -----
 0   itching                       1440 non-null   int64
 1   skin_rash                     1440 non-null   int64
 2   continuous_sneezing           1440 non-null   int64
 3   shivering                     1440 non-null   int64
 4   chills                        1440 non-null   int64
 5   joint_pain                    1440 non-null   int64
 6   stomach_pain                  1440 non-null   int64
 7   acidity                       1440 non-null   int64
 8   ulcers_on_tongue              1440 non-null   int64
 9   muscle_wasting                1440 non-null   int64
 10  vomiting                      1440 non-null   int64
 11  fatigue                       1440 non-null   int64
 12  weight_loss                   1440 non-null   int64
 13  restlessness                  144

###### Data has 58 columns. The 57 of the columns are the symptoms, encoded as integer data, and the last column 'prognosis' is the target or categorical data for disease labels

In [8]:
train_data.isnull().any()

itching                         False
skin_rash                       False
continuous_sneezing             False
shivering                       False
chills                          False
joint_pain                      False
stomach_pain                    False
acidity                         False
ulcers_on_tongue                False
muscle_wasting                  False
vomiting                        False
fatigue                         False
weight_loss                     False
restlessness                    False
lethargy                        False
patches_in_throat               False
irregular_sugar_level           False
cough                           False
high_fever                      False
breathlessness                  False
sweating                        False
indigestion                     False
headache                        False
nausea                          False
loss_of_appetite                False
pain_behind_the_eyes            False
back_pain   

In [9]:
train_data

itching  skin_rash  continuous_sneezing  shivering  chills  joint_pain  \
0           0          1                    0          0       0           0   
1           0          1                    0          0       0           0   
2           0          0                    0          0       0           0   
3           0          1                    0          0       0           0   
4           0          1                    0          0       0           0   
...       ...        ...                  ...        ...     ...         ...   
1435        0          0                    0          0       0           0   
1436        0          0                    0          0       0           0   
1437        0          0                    0          0       0           0   
1438        0          0                    0          0       0           0   
1439        0          0                    0          0       0           0   

      stomach_pain  acidity  ulcers_on_tongue  muscle_wasting  ...  \
0                0        0                 0               0  ...   
1                0        0                 0               0  ...   
2                0        0                 0               0  ...   
3                0        0                 0               0  ...   
4                0        0                 0               0  ...   
...            ...      ...               ...             ...  ...   
1435             0        1                 0               0  ...   
1436             0        1                 0               0  ...   
1437             0        1                 0               0  ...   
1438             0        1                 0               0  ...   
1439             0        1                 0               0  ...   

      red_spots_over_body  watering_from_eyes  increased_appetite  polyuria  \
0                       0                   0                   0         0   
1                       0                   0                   0         0   
2                       0                   0                   0         0   
3                       0                   0                   0         0   
4                       0                   0                   0         0   
...                   ...                 ...                 ...       ...   
1435                    0                   0                   0         0   
1436                    0                   0                   0         0   
1437                    0                   0                   0         0   
1438                    0                   0                   0         0   
1439                    0                   0                   0         0   

      lack_of_concentration  visual_disturbances  pus_filled_pimples  \
0                         0                    0                   1   
1                         0                    0                   1   
2                         0                    0                   1   
3                         0                    0                   0   
4                         0                    0                   1   
...                     ...                  ...                 ...   
1435                      0                    1                   0   
1436                      0                    1                   0   
1437                      0                    1                   0   
1438                      0                    1                   0   
1439                      0                    1                   0   

      blackheads  scurring  prognosis  
0              1         1          0  
1              1         1          0  
2              1         1          0  
3              1         1          0  
4              0         1          0  
...          ...       ...        ...  
1435           0         0         11  
1436           0         0         11  
1437           0         0    

#### Create Test and Train Data

In [12]:
X_train = train_data.iloc[:, :-1]
X_test = test_data.iloc[:,:-1]

In [14]:
y_train = train_data.iloc[:, -1:]
y_test = test_data.iloc[:, -1:]

In [16]:
y_train.shape
X_train.shape

(1440, 57)

- X_train data has about 57 features and 1440 observation unit that means our input matrix has 1440 rows and 57 columns for training

- y_train has about 1 features and 1440 observation unit that means our input matrix has 1440 rows and 1 columns dor training

In [17]:
print("X test shape: ",X_test.shape)
print("y test shape: ",y_test.shape)

X test shape:  (12, 57)
y test shape:  (12, 1)


##### Featuring Scaling for Normalization

In [20]:
x = np.array([1,23,5,564,56,876,7,-123])

standardized_X = (x - np.mean(x))/np.std(x)
normalized_X = (x-np.min(x) / np.max(x) - np.min(x))
print('Standardized array : ', standardized_X)
print('normalized array : ', normalized_X)

Standardized array :  [-0.53531619 -0.46806733 -0.52308913  1.18564322 -0.36719405  2.13935429
 -0.5169756  -0.91435521]
normalized array :  [1.24140411e+02 1.46140411e+02 1.28140411e+02 6.87140411e+02
 1.79140411e+02 9.99140411e+02 1.30140411e+02 1.40410959e-01]


In [22]:
# With sklearn scalers

from sklearn.preprocessing import StandardScaler
stds = StandardScaler()
x = x.reshape(-1,1) # for sklearn method. (Use 2 dimensional vectors)

x = stds.fit_transform(x)
x

array([[-0.53531619],
       [-0.46806733],
       [-0.52308913],
       [ 1.18564322],
       [-0.36719405],
       [ 2.13935429],
       [-0.5169756 ],
       [-0.91435521]])

##### Model Building 

Implementing Basic Example ANN Structure with OOP


Reference: https://www.geeksforgeeks.org/implementing-ann-training-process-in-python/

In [23]:
class NeuralNet(object):
    def __init__(self):
        # Generate random numbers
        np.random.seed(1)
        
        # Assign random weights to a 3 * 1 matrix
        self.synaptic_weights = 2 * np.random.random((3, 1)) - 1
        
    # The sigmoid function method
    def _sigmoid(self,x):
        return 1 / (1 + np.exp(-x))
    
    # Derivative sigmoid
    def derivative_sigmoid(self, x):
        return x * (1 - x)
    
    # Train the neural network and adjust the weights each time
    def train(self, inputs, outputs, iteration_number):
        for iteration in range(iteration_number):     
        
            # Pass the training set through network
            output = self.learn(inputs)
        
            # Calculate the error
            error = outputs - output
        
            # Adjust the weights by a factor
            factor = np.dot(inputs.T, error * self.derivative_sigmoid(output))
            self.synaptic_weights += factor
        
    # calculate z    
    def learn(self, test_inputs):
        return self._sigmoid(np.dot(test_inputs, self.synaptic_weights))

###### Fitting Model and Prediction

In [25]:
# Initialize

neural_net = NeuralNet()

# The training set
inputs = np.array([[0, 1, 1], [1, 0, 0], [1, 0, 1]])
outputs = np.array([[1, 0, 1]]).T

# train the neural network
neural_net.train(inputs, outputs, 50)

test_inputs = np.array([1, 0, 1])
threshold = 0.5
if neural_net.learn(test_inputs) >= threshold:
    print("Our test example output is: 1")
else:
    print("Our test example output is: 0")

Our test example output is: 1


In [26]:
# transform into dummies for y_train (prognosis variable)
y_train_dum = pd.get_dummies(y_train)
y_train_dum

prognosis
0             0
1             0
2             0
3             0
4             0
...         ...
1435         11
1436         11
1437         11
1438         11
1439         11

[1440 rows x 1 columns]





##### Building ANN Structure with Keras Library


In [28]:
!pip3 install tensorflow

^C


In [30]:
# import tensorflow and keras

import tensorflow as tf

ModuleNotFoundError: No module named 'tensorflow'